### ISSUE TITLE
[Python] Clarify meaning of `-py3` and python2 vs. python3 compatibility

### ISSUE THREAD

**Issue by m7thon** [swig/swig#695]
Currently, according to the SWIG docs:
- `swig -python` will generate wrappers that should work with python2 _and_ python3, and
- the `-py3` flag can be specified to enable python3 specific features which might break backwards compatibility with python2 (this clearly was the original intention of the `-py3` flag).
- The swig python examples (in the docs) make little/no mention of the `-py3` flag.

However, the SWIG code assumes in several places that **the `-py3` flag must be given to enable python3 compatibility**, i.e., that SWIG cannot generate wrapper code that is simultaneously compatible with python2 and python3. Furthermore, all SWIG tests are run with the `-py3` flag when generating wrapper code for python3.

This is a discrepancy asking for trouble, and should be resolved either by **(1) clarifying that python3 requires `-py3`** in the documentation (and possibly merging the `-python` and `-py3` flags), or by **(2) making `swig -python` work for python2 _and_ python3** by fixing the codebase. I list some things worth considering that I came across for anyone interested in tackling this 

**EDIT: This is now resolved via Option 2 below.**
## Option 1: Python3 really does require the `-py3` flag
- Obviously, the documentation under "Python 3 Compatibility" needs to be updated. But also, the typical use cases throughout the documentation should show the use of `-py3` (perhaps by default?) to make sure people are aware of this flag.
- The logic of having `-python` as a language specifier that really means `-python2`, and having `-py3` as an extra flag that really means `-python3`, is unintuitive and unsymmetrical. It would make more sense to use either language specifiers `-python2` and `-python3`, or one language specifier `-python` and flags `-py2` and `-py3` to select the language version, perhaps defaulting to `-py2`. The typical usage of `swig -python -py2` could be shown throughout the documentation to emphasized the point that `-python -py2` wrapped code is not guaranteed to work under Python 3.
- The minor downside of this is that any project that wishes to include pre-generated python wrapper code and wishes to support python2 and python3 is forced to handle both cases separately.
- Python distutils and setuptools support swig-wrapped extensions by automatically swigging and building them. They don't ever specify the `-py3` flag, so this needs to be fixed.
## Option 2: `-python` means Python 2 _and_ 3 compatibility

There are only three code-paths that currently use the `-py3` flag for Python 3 compatibility. These are:
- [x] Code activated by the `-relativeimports` flag for relative module imports (without `-relativeimports` everything is fine for Python 2 and 3). This scared me, so I passed it over. However, selecting the correct import syntax can definitely be done in python code at runtime without relying on `-py3`.
  
  Fixed by #698
- [x] Code to convert **octal** default argument values to python default values (since Python 2 uses eg. `012` while Python 3 uses `0o12`). This could be resolved by converting the octal to a different representation, eg. hexadecimal, or alternatively by using a new python function, say `_swig_octal(oct_string)` that parses the value string in a python version dependent way, and then writing the default argument value, say `0o12`, as `_swig_octal('0o12')`.
  
  Fixed by #707
- [x] The division operator mappings
      - `*::operator/` to `__div__` or `__truediv__`
      - `*::operator /=` to `__idiv__` or `__itruediv__`
  Note that Python is perfectly happy to have both defined and calling the same `C++` operator. This would ideally call for some `%alias` mechanism similar to `%rename` but allowing multiple new names.
  
  Fixed by #709
- [x] Also, the tests would need modifications to run the combination of `swig -python` wrapped code with Python 3.
  
  This is partially achieved by commit 7a26e33.
## Afterthought:

This issue is not pressing, since problems can _only_ happen when using `swig -python` wrapper code under Python 3 _and_ using one of the cases listed above under Option 2. So perhaps it makes sense to think about it and leave it for some new major SWIG version to fix.

**Comment by vadz** [comment]
FWIW I'm (another?) one of those people who were implicitly sure that SWIG was using the first option before the current discussion started and I still don't really understand what `-py3` would mean if this were not the case.

However if it's possible to change the generated code to be Python-version-agnostic and thus remove the need for `-py3` entirely, it would, of course, be even better. It does risk making the generated C code even messier, but it's probably still worth it.

**Comment by mromberg** [comment]
My 2 cents.   In some ways python2 and python3 are two different languages.  The later is not backwards compatible with the former.   While there is a need to support both for "the transition", the transition has been going on now for almost a decade).  How long should swig be expected to support the strange language that is both python2 and python3?

  It makes a lot of sense to have the two flags suggested (-python2 and python3).  Then treat them as separate languages.  Of course the internals of swig could share quite a bit of code.  But the output would be for one language or the other.   Supporting both (which stretches from 2.4 - 3.6) covers a wide range of python changes.

  I've read that big things are planned for swig in the 3.1 release regarding python.   So, this might be a good time to more clearly make a distinction with the command line switches as well.

**Comment by wsfulton** [comment]
I've committed df30ae0f0b7ddf6755cdbe776b9073aee4ec78e5 which makes it easy to test python3 without passing the -py3 option to swig as follows:
```
make check-python-test-suite PY3=1 SWIGOPTPY3=
make check-python-examples PY3=1 SWIGOPTPY3=
```
I ran this using python3.5 on Ubuntu and there were just 3 test-suite failures:
```
Traceback (most recent call last):
  File "operator_overload_runme3.py", line 6, in <module>
    pop = Op(6)/Op(3)
TypeError: unsupported operand type(s) for /: 'Op' and 'Op'

Traceback (most recent call last):
  File "default_args_runme3.py", line 158, in <module>
    run('default_args')
  File "default_args_runme3.py", line 10, in run
    default_args = __import__(module_name)
  File "/home/william/swig/github/swig/Examples/test-suite/python/default_args.py", line 110
    def trickyvalue3(first, mode=0644):
                                    ^
SyntaxError: invalid token

Traceback (most recent call last):
  File "python_abstractbase_runme3.py", line 9, in <module>
    assert issubclass(Mapii, MutableMapping)
AssertionError
```
These correlate with @m7thon's limitations in Option2 above.
when adding in SWIG_FEATURES=-builtin, the entire test-suite passed.

The import_packages example failed though with and without -builtin. So looks like the documentation is mostly right and -py3 is mostly not required for Python3. I was also under the impression that -py3 had to be used for Python3 so sorry for confusion. Given the current state, it looks like that -py3 is _recommended_ rather than _absolutely required_.

**Comment by wsfulton** [comment]
Now running:
```
~/swig/github/swig/Examples/python/import_packages (master=)$ make -s clean check PY3=1 SWIGOPTPY3=
Testing same_modnames1 - %module(package=...) + python 'import' in __init__.py
  Finished importing pkg2.foo
  Successfully created object pkg2.foo.Pkg2_Foo
Testing same_modnames2 - %module(package=...) + python 'import' in __init__.py
  Finished importing pkg1.pkg2.foo
  Successfully created object pkg1.pkg2.foo.Pkg2_Foo
Testing from_init1 - %module(package=...) + python 'import' in __init__.py
Traceback (most recent call last):
  File "runme3.py", line 12, in <module>
    import py3.pkg2
  File "/home/william/swig/github/swig/Examples/python/import_packages/from_init1/py3/pkg2/__init__.py", line 1, in <module>
    from .bar import Pkg2_Bar
  File "/home/william/swig/github/swig/Examples/python/import_packages/from_init1/py3/pkg2/bar.py", line 101, in <module>
    import foo
ImportError: No module named 'foo'
```
Same for -builtin.

@mromberg or @m7thon would you mind looking at this too? I havn't check that this is not a regression and @m7thon mentioned above that ```-relativeimports``` could be fixed at runtime too? 

Some of the runme.py files don't have any print statements in them, which would be a good enhancement too in order to follow what is going on.

**Comment by m7thon** [comment]
I took a look at the failing results from
```sh
make check-python-test-suite PY3=1 SWIGOPTPY3=
make check-python-examples PY3=1 SWIGOPTPY3=
```

and all fall into the cases I listed above under "Option 2". The only noteworthy cases are

* The `python_abstractbase_runme3.py` failure, but this checks a specific Python 3 feature (abstract base classes) that is obviously only enabled if the `-py3` flag is given. This test should therefore not be run if the `-py3` flag was not given. (It is also not run for Python 2 at all).

    So this is an artifact of the test suite not being designed for the combination of `swig -python` with Python 3.

* The `import_packages` examples failures, but even though the names are misleading, *all* of the failing examples add the `-relativeimports` flag in their Makefiles, so this is really due to `-relativeimports`.

**Comment by m7thon** [comment]
Note that the current division operator mapping to `__div__` is actually broken under Python 2 as well if a user uses `from __future__ import division` in their own script (which is actually quite common to do).

**Comment by mromberg** [comment]
This patch:
https://github.com/swig/swig/compare/master...mromberg:relative

  Seems to fix the failures with 

make check-python-examples PY3=1 SWIGOPTPY3=

  There appears to be some documentation generation stuff that depends on py3 as well.  Don't know how that affects folks who desire to run swig once and use the .py file for both python2 and python3:
bool funcanno = py3 ? true : false;
    String *params = NewString("");
    String *_params = make_autodocParmList(n, false, is_calling, funcanno);

**Comment by wsfulton** [comment]
Patch looks okay, thanks. Could you create a pull request and add in the appropriate corrections to the docs in http://swig.org/Doc3.0/Python.html#Python_absrelimports too please?

**Comment by m7thon** [comment]
Your "fix" for the `-relativeimports` stuff is basically ok, but I would do it more carefully/precisely in the following way:

* Don't import `sys` (you don't want to pollute the module namespace or clobber any users symbol). But note that `version_info` was already imported (and deleted) at the top of the shadow module.
* Therefore, move the "del version_info" to the end of the shadow module and use it throughout.
* But, to be extra careful that this does not conflict any users symbol, rename `version_info` to `_swig_python_version_info` throughout (i.e., use `from sys import version_info as _swig_python_version_info`).
* Use the Python 3 syntax *only* for version >= 3.0.0 (you use ">" and 2.7.0), just to make sure the new code is really equivalent to the old -- you never know :-)

**Comment by m7thon** [comment]
Another note about the import stuff: The docs mention in (36.11.3 Enforcing absolute import semantics)[http://www.swig.org/Doc3.0/SWIGDocumentation.html#Python_absimport] that one can enforce absolute import semantics (which makes the syntax stricter) for Python 2.5 to 2.7 by using
```python
%pythonbegin %{
from __future__ import absolute_import
%}
```
This breaks `-relativeimports` on Python 2, so perhaps this should be fixed while we are at it. We can do this in an elegant way:

At the top of the shadow module, after importing `version_info`, insert:
```python
try:
    if type(absolute_import) is __future__._Feature:
        _swig_absolute_import = True
except:
    if version_info >= (3,0,0):
        _swig_absolute_import = True
    else:
        _swig_absolute_import = False
```

Then, use as condition for which import syntax to use simply `if _swig_absolute_import:`. The `version_info` is then not needed throughout the shadow module.

**Comment by m7thon** [comment]
Oh, and the function annotations are a new feature of Python 3 (like abstract base classes) that needs to be enabled by `-py3`, so that is not an issue here.

**Comment by wsfulton** [comment]
2d5c32048c324e73c94edf757e46b6491f92a362 ensures the python_abstractbase test is only run when -py3 is used, so this one will always pass now. 

On another note, this test isn't run for -builtin and I don't recall why I added in a comment in python_abstractbase_runme3.py:
```
 # Builtin types can't inherit from pure-python abstract bases
```
It seems to be a rather big limitation.

**Comment by wsfulton** [comment]
It is usually better to avoid the catch all handler, so change
```
except:
```
to
```
except NameError:
```
I don't see how we can adding in ```from __future__ import absolute_import``` as it must be the first line, so we can't test for version < 2.5 in the runme.py file. Unless there are any other ideas, we'll skip testing this aspect in the test-suite.

**Comment by mromberg** [comment]
> there is nothing special about version 2.7.0

  It has been a while (and my memory is not what it used to be).  But I thought the reason python 2.7 was created in the first place was to be a transitional version of python between 2 and 3:  

https://docs.python.org/3/howto/pyporting.html

  Given that, it makes sense (in my mind) for swig to use it as the point where python (generated by swig) should start using python 3 features (that are supported by 2.7+).

**Comment by mromberg** [comment]
> Don't import sys (you don't want to pollute the module namespace or clobber any users symbol). But note that version_info was already imported (and deleted) at the top of the shadow module.

  Yea.  I know that version_info is used in other places.  But I'm not exactly sure where they all are.  And I'm not sure all the places where rel_import_directive_string() is being called from.   It would be nice to clean up the import of sys so that it is done once and all that.  But these bits of python code are sprinkled around a bit.  So, it is not clear to me where one would get version_info so that it was done before any swig code would need it and exactly where it should be deleted.  Importing sys when it was needed for sure seems a good way to K.I.S.S.   We already know there are gaps in the swig test coverage.  :)

**Comment by mromberg** [comment]
@m7thon Do you happen to know if the following is valid python syntax in all the versions of python supported by swig?

from sys import version_info as _swig_version_info

  If so, then we could/should use something along those lines for the python runtime version check.

**Comment by mromberg** [comment]
> Code to convert octal default argument values to python default values (since Python 2 uses eg. 012 while Python 3 uses 0o12). This could be resolved by converting the octal to a different representation, eg. hexadecimal, or alternatively by using a new python function, say _swig_octal(oct_string) that parses the value string in a python version dependent way, and then writing the default argument value, say 0o12, as _swig_octal('0o12').

  I looked at this a bit.  And (unless I'm missing something) there does not appear do be an easy/good way to use octal integer literals with swig and expect it to create code which is at the same time both valid in python2 and python3.

  Right now convertValue(String *v, SwigType *type) is using py3 to do this (at swig time).   Moving it (or the chunk of it that deals with python octal syntax) to runtime would have messy complications:

  The function (convertValue) is documented as returning Nil if the string can not be a python literal or constant.   Deferring this to runtime for octal numbers seems messy.   I'm assuming that convertValue is used by swig quite often to convert strings it has parsed.  I would guess that its use is not isolated to an easily testable area.

   My gut tells me that if your wrapper uses octal numbers and you expect the swig wrapper to work correctly with them then you *must* generate the wrappers with the -py3 switch for python3 and without it for python2.  It is a fundamental way python3 is not backwards compatible with python2.

  An alternative would be to detect when C style octal literals are found convert them to hex or decimal literals.   This would work with both pythons and not need the -py3 switch.  But I can see folks complaining about this as many octal numbers are used (and memorized in octal format) for things like file permissions.  So, the expectation may be there for swig to keep the same notation.

**Comment by m7thon** [comment]
`from sys import version_info as _swig_version_info` is valid for all versions of python 2 [[ref](https://docs.python.org/release/2.0/ref/import.html)]

You can put the above code snippet for `_swig_absolute_import`, or, if you don't feel like fixing that, then instead `from sys import version_info as _swig_python_version_info` between [these](https://github.com/swig/swig/blob/1495313e53c26f6e0394e0c7f0411fc27f1054ce/Source/Modules/python.cxx#L1050-L1051) lines.

To put the `del _swig_python_version` or `del _swig_absolute_import` statement at the end of the file, print it [here](https://github.com/swig/swig/blob/1495313e53c26f6e0394e0c7f0411fc27f1054ce/Source/Modules/python.cxx#L1054).

However, please do **not** use either of `import sys`, or `from sys import version_info`, since it will overwrite any previously defined symbol called "sys" or "version_info" (I can give you a test case that breaks, if you like).

**Comment by m7thon** [comment]
I'll do a fix for the octal thing. Hang on...

**Comment by mromberg** [comment]
Here is another area where the whole idea of code that works for python2 and python3 breaks down.  Metaclasses.  Swig uses them to support the "feature:python:nondynamic" flag.  This feature is currently broken in swig-3.0.X while using -O.  Part of the reason it is broken is because python3 does not do anything with the __metaclass__ attribute and the swig code to enforce nondynamic is not being run.

  I have a fix for nondynamic and -O.  But, the classes swig wraps which this flag need to have a metaclass attached to them.  There are two ways to do this:

Python2:

class A(object):
    __metaclass__ = MetaClass
    ...

  Python3:

class A(object, metaclass=MetaClass):
    ...

  Python3 will happily ignore __metaclass__ syntax of python2 which is fine except that it also does not use the MetaClass and fails the tests.  And of course python2 chokes hard on the python3 syntax.   

  Without relying on -py3 at swig time, I see the options as:

  1)  Punt the nondynamic feature.

   2) Don't use metaclasses to implement them (not exactly sure what other method would work).

   3) Have the swig wrapper code use some external library like six

   4)  Have the wrapper code implement its own version of the six code to handle this.

  Relying on the -py3 switch and just spitting out the correct code for either python2 or python3 (but not both) seems pretty attractive.

**Comment by m7thon** [comment]
I created a new issue #719 to discuss the nondynamic stuff. As you can see, my vote is strongly for your option 1), i.e., get rid of it.

**Comment by arpitn30** [comment]
Sorry, this might be the wrong place to ask this question but I couldn't get a satisfactory answer anywhere else.
In order to support both Python 2 and Python 3, is there any "sys.version_info.major" type function for C++ or will we have to operate in the grey area writing only the code that works in both of the python versions?

**Comment by mromberg** [comment]
@arpitn30 The python C API https://docs.python.org/3/c-api/index.html has pre-processor macros you can use to find the python version numbers.  for example:

#if PY_MAJOR_VERSION < 3
// Python2 code
#else
// Python3 code
#endif

  For a swig wrapper you will need to do this at compile time as you can't compile and link a swig module using both python2 and python3 at the same time.  The swig module could contain code such as the above to detect the python version it is being built with.  Then one could build two separate swig modules.  One for python2 and one for python3.

**Comment by wsfulton** [comment]
Regarding metaclasses and "feature:python:nondynamic" with -O and the comment:

> 4. Have the wrapper code implement its own version of the six code to handle this.

This is exactly what is now in place, see #718. When using just -python the generated code is compatible with python 2 and python 3 using a slimmed down version of` six.add_metaclass`. When using -py3, the Python 3 syntax for metaclasses is output instead.

**Comment by wsfulton** [comment]
I think it is clear now in this issue that -py3 generates Python 3 syntax that is not compatible with Python 2. Unless someone thinks this is not clear enough in the Python.html documentation, I'll close this issue.

**Comment by wsfulton** [comment]
Closing as mentioned recently.
